### PR TITLE
Add --report option to report state

### PR DESCRIPTION
**Pull request by stdweird**

### CODE REVIEW

File: src/main/perl/CLI.pm [hunk 1]
```diff
@@ -387,20 +398,134 @@ sub check_options
...
-    if ($self->option('configure') && $self->option('unconfigure')) {
-        $self->error('configure and unconfigure options cannot be used simultaneously');
+    my $report_format = $self->option('report-format');
+    if (!grep {$_ eq $report_format} @SUPPORTED_REPORT_FORMATS) {
+        $self->error("Unsupported report format $report_format, use of one ".join(', ', @SUPPORTED_REPORT_FORMATS));
         $self->finish(-1);
     }
 }
 
+# simple wrapper around _print for unittesting
```
Comment (ned21) [inline src/main/perl/CLI.pm]: s/_print/print/ ?

File: src/main/perl/CLI.pm [hunk 2]
```diff
@@ -387,20 +398,134 @@ sub check_options
...
+    my $perfdata = "failed=$nr";
+    if (defined($cfgname)) {
+        $perfdata .= ", version=$cfgname";
+    }
+    my ($ec, $msg);
+    if ($nr) {
+        my @comps = sort keys %$state;
+        $msg = "ERROR $nr components with error: ".join(',', @comps);
+        my $size = length($msg) + length($perfdata) + 3; # 3 for separator ' | '
+        if ($size > 255) {  # TODO: double check if this is inlcuding the perfdata (let's assume it is)
```
Comment (ned21) [inline src/main/perl/CLI.pm]: Address and/or remove TODO

File: src/main/perl/ComponentProxyList.pm [hunk 3]
```diff
@@ -808,6 +812,61 @@ sub set_state
 }
 
 
+=item get_states
+
+Gather the state information of the components in C<statedir>.
+
+Return hashref with component name as key and state information:
```
[the remark this answers is not in the document]
Comment (stdweird) [reply src/main/perl/ComponentProxyList.pm]: statefiles contain error messages. presence of file indicates error (afaik)

**Review by jrha — APPROVED** [review]

**Review by ned21 — APPROVED** [review]

**Comment by jrha** [comment]
@stdweird since this was merged nightlies have been failing on CentOS6 and CentOS7 (but passing on CentOS8) and I cannot figure out why the behaviour is different.

Specifically the 'failed ouch component' test:
```
…
INFO: /var/run/quattor-components/ouch status is mode=420
INFO: /var/run/quattor-components/woohaa status is mode=420
[
  [
    '2 components with error'
  ],
  [
    '  woohaa failed on Thu Nov 29 21:33:09 1973 with message woopsie'
  ],
  [
    ${$VAR1->[1][0]}
  ]
]

  Failed test 'failed ouch component'
  at src/test/perl/cli.t line 175.
                  '  woohaa failed on Thu Nov 29 21:33:09 1973 with message woopsie'
    doesn't match '(?^:^  ouch failed on .*? 2001 (no message))'
…
```

**Comment by jrha** [comment]
I thought it might be an ordering issue, but when I tried switching the indices in the tests it apparently had no effect.

**Comment by jrha** [comment]
Sure, they're also available here:
http://yum.quattor.org/nightlies/2021-06-07T00/CentOS-6/log.gz
http://yum.quattor.org/nightlies/2021-06-07T01/CentOS-7/log.gz
http://yum.quattor.org/nightlies/2021-06-07T02/CentOS-8/log.gz

**Pull request merged**